src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,184 @@
+    _scan_reference_segment<T>(reference_segment);
+  } else {
+    Fail("Segment type is not supported.");
+  }
+}
+
+template <typename T>
+void TableScan::_scan_value_segment(const ChunkID chunk_id, const std::shared_ptr<ValueSegment<T>>& segment) {
+  if (variant_is_null(search_value())) {
+    _scan_for_null_value<ValueSegment<T>, T>(chunk_id, segment);
+    return;
+  }
+  const auto selector = Selector<T>(scan_type(), type_cast<T>(search_value()));
+  const auto segment_size = segment->size();
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < segment_size; ++chunk_offset) {
+    const auto value = segment->get_typed_value(chunk_offset);
+    // We don't want to match NULL, regardless of the condition.
+    if (value && selector.selects(value.value())) {
+      _emit(chunk_id, chunk_offset);
+    }
+  }
+}
+
+template <typename T>
+void TableScan::_scan_dictionary_segment(const ChunkID chunk_id, const std::shared_ptr<DictionarySegment<T>>& segment) {
+  if (variant_is_null(search_value())) {
+    _scan_for_null_value<DictionarySegment<T>, T>(chunk_id, segment);
+    return;
+  }
+  const auto value_ids = segment->attribute_vector();
+
+  const auto typed_search_value = type_cast<T>(search_value());
+  const auto lower_bound_value_id = segment->lower_bound(typed_search_value);
+  const auto upper_bound_value_id = segment->upper_bound(typed_search_value);
+
+  const auto emit_callback = [this, &chunk_id](const ChunkOffset chunk_offset) { _emit(chunk_id, chunk_offset); };
+
+  const auto selector =
+      DictionarySegmentSelector(scan_type(), search_value(), emit_callback, value_ids, segment->null_value_id());
+  selector.select(lower_bound_value_id, upper_bound_value_id);
+}
+
+template <typename T>
+void TableScan::_scan_reference_segment(const std::shared_ptr<ReferenceSegment>& segment) {
+  const auto segment_size = segment->size();
+  const auto pos_list = segment->pos_list();
+
+  if (variant_is_null(search_value())) {
+    _scan_for_null_value<ReferenceSegment, T>(ChunkID{0}, segment, pos_list);
+    return;
+  }
+
+  const auto selector = Selector<T>(scan_type(), type_cast<T>(search_value()));
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < segment_size; ++chunk_offset) {
+    const auto value = segment->_get_typed_value<T>(chunk_offset);
+    // We don't want to match NULL, regardless of the condition.
+    if (value && selector.selects(value.value())) {
+      // Emit using the existing PosList.
+      // This way, we are able to omit ReferenceSegments referencing ReferenceSegments.
+      _emit(pos_list->operator[](chunk_offset));
+    }
+  }
+}
+
+template <typename SegmentType, typename T>
+void TableScan::_scan_for_null_value(const ChunkID chunk_id, const std::shared_ptr<SegmentType>& segment,
+                                     const std::shared_ptr<const PosList>& pos_list) {
+  if (scan_type() != ScanType::OpNotEquals) {
```

Comment:
The implementation detail, that when scanning with NotEquals, 'Null NotEquals any other value' evaluates to true could probably be mentioned somewhere in a broad operator description. (At least I believe this is not standard SQL)